### ISSUE TITLE
DownloadCentral stops permanently without any error messages when a download times out on Mac OS X (Stuck on (large) Download Central Job (in semaphore or /bin/bash ?))

### ISSUE THREAD

**Issue by markusmaresch** [gnosygnu/xowa#80]
Repeatedly stuck on larger download central job on MAC/OSX - like: "Deutsch Wikipedia - Images (2016-06) [42.86 GB]", but probably also for other larger packages.

The process just stops - there is no progress in the tmp/bldr/*/download.tmp file, which usually is.
Terminating the process allows a continuation - seems ok so far.

OSX is providing a crash-report - it seems to be waiting on some semaphore, or /bin/bash crashed, I can't interpret this right now.

I can try help debugging ..
[xowa_stuck_downloadcentral.zip](https://github.com/gnosygnu/xowa/files/430377/xowa_stuck_downloadcentral.zip)

Great feature this download central !! So much easier than before :-) !!

**Comment by gnosygnu** [comment]
Hi! Thanks for the ticket as well as the debug dump!

Unfortunately, I wasn't able to reproduce it now. I tried it on a Mac OS X 10.10.5 (Yosemite) and was able to download both German Wikipedia Articles and Images (it took about 9 hours).

Also, the debug dump doesn't offer any clues. It looks like the OS killed the task ("process was unresponsive for 19 seconds before sampling"), but I'm not sure why. XOWA is downloading constantly, so it shouldn't be stalled 

Let me look at it a little further. Out of curiosity, where are you downloading from? I'm on the East Coast of the United States for comparison. I'm just wondering if there is some lag in downloading which is causing OS X to mark XOWA for termination.

> The process just stops - there is no progress in the tmp/bldr/*/download.tmp file, which usually is. 
> Terminating the process allows a continuation - seems ok so far.

Yup, terminating and continuation will work. I designed it to handle abrupt terminations (necessary for Android), so you can continue going that route.

**Comment by markusmaresch** [comment]
Hi - additional feedback.
I am in Europe, and can download at about 7 MB/sec - German or English takes hours, yes - but it basically works. When SWT on MacOS is stuck, I terminate it the soft way (close window), then kill it, because it is not responding. After restart it just continues - well.
So - I could live with it, as I know how to handle it.
Up to you if you want to suspend this ticket :-) 

Related: I did get a HASH error, which Xowa handled gracefully by downloading the particular package again - but this was only once and maybe just a coincidence.  Still puzzled as to why this would happen these days - may be caused by statistics, with 1 in 10^12..10^15, things can happen..

Is there an easy way to attach to the Java VM when it is stuck - in order to find out more ? (just happened again as I type this..)

**Comment by gnosygnu** [comment]
> I am in Europe, and can download at about 7 MB/sec - German or English takes hours, yes - but it basically works.

7 MB / sec is actually not bad. Sometimes I get spurts of 500 kb here in the East Coast US. :)

> When SWT on MacOS is stuck, I terminate it the soft way (close window), then kill it, because it is not responding. After restart it just continues - well.
> So - I could live with it, as I know how to handle it.
> Up to you if you want to suspend this ticket :-) 

Nope. I'm not planning on suspending the ticket. I just don't have don't have any credible clues to go on.

> Related: I did get a HASH error, which Xowa handled gracefully by downloading the particular package again - but this was only once and maybe just a coincidence. Still puzzled as to why this would happen these days - may be caused by statistics, with 1 in 10^12..10^15, things can happen.

Yeah, I try to make sure I verify everything in Download Central, so if anything goes wrong, it will restart. That said, I usually don't get errors on the desktop side, so your situation is strange

> Is there an easy way to attach to the Java VM when it is stuck - in order to find out more ? (just happened again as I type this..)

Not that I know of. I think the problem is in the OS level though, so attaching to the VM probably won't yield any clues.

Right now, I'm really confused about why it raises a "process unresponsive error". Download Central does 2 things:
* Runs the download in a separate thread (so the SWT GUI should never be unresponsive)
* Updates progress every second. (so even if it was the main UI thread, it wouldn't be unresponsive for more than a second)

Let me put together a debugging build for this weekend. It'll have a switch that logs every download central message / action.  Hopefully, it'll pinpoint exactly where the problem is.

**Comment by markusmaresch** [comment]
I can help debugging this with the right build as soon as needed. :-)

**Comment by gnosygnu** [comment]
Cool. I'm delaying the release for one day so I can get some other things in there. I should be posting again here tomorrow night when it's out.

Thanks!

**Comment by gnosygnu** [comment]
I added some additional logging to tonight's release. When you get a chance, please try the following:

* Download the latest release: https://github.com/gnosygnu/xowa/releases/tag/v3.8.5.1
* Go to "home/wiki/Options/Config_script"
* Enter <code>xowa.api.addon.bldr.central.log_verbose = 'y';</code>
* Press "Save"
* Go to Download Central and download one of the German wikis. For what it's worth, I uploaded a new version of 2016-08 German Wikipedia today (with an image update)
* While its downloading, check "home/wiki/Special:XowaSystemData?type=log_session". You will see entries at the bottom like "20160830_030900.974 xobc:notify: task_id=1 cmd_id=0 prog_data_cur=272736 prog_data_end=49931717 rate_cur=259501.42721217888"
* When it fails again...
    * Post the log file here. Note that the path for the log file is listed at the top of the page "Path: ..."
    * Post the crash dump here again.

Between the two files, I'm hoping that I see how "unresponsive" the GUI was. I'm also hoping that the log file shows whether or not there were long spans of unresponsiveness. That is, the crash happened at 19 seconds; theoretically, there should be smaller gaps before then.

Let me know if you run into any issues with the debug setup. Thanks again for all your assistance.

**Comment by markusmaresch** [comment]
Feedback:
I successfully downloaded the updated Germans with debug on 3.8.5.1 on OS-X
It got stuck once - yet this does not seem to be reflected in the log.
What I did: 
Once it got stuck, I tried to hit the buttons top and play - without effect, they were not responding. 
The GUI was not entirely frozen - the green progress bar responded nicely (rescale in x) to a resize of the entire application window. THEN I pushed the green forward and back, because I wanted to look at the session log. Going back triggered a continuation - the freeze was in the 003.zip
Conclusion: Occasionally it is still stuck, the GUI is responsive, and playing with the GUI, a restart and continuation can be triggered.

Thanks for the work on this :-) 

[20160830_084100.log.zip](https://github.com/gnosygnu/xowa/files/444605/20160830_084100.log.zip)

**Comment by gnosygnu** [comment]
Hey, thanks for the follow-up

> I successfully downloaded the updated Germans with debug on 3.8.5.1 on OS-X

Cool. :+1: 

> It got stuck once - yet this does not seem to be reflected in the log.

Yeah, actually there's another log (sorry there are two. the 2nd one is primarily for Android)

The one you sent me was from: /Users/markus/xowa/user/anonymous/app/tmp/xolog

The one that should have more debug info is in: /Users/markus/xowa/user/anonymous/app/tmp/log

If you still have it, send it on over. If not, can you try triggering another download (like Simple Wikipedia) and seeing if it fails?

> Once it got stuck, I tried to hit the buttons top and play - without effect, they were not responding.
> The GUI was not entirely frozen - the green progress bar responded nicely (rescale in x) to a resize of the entire application window.

Thanks. That description is very helpful. It tells me:
* the HTML browser is not responsive to user input ("clicking top and play didn't do anything")
* the HTML browser was still responsive to the program ("the green progress bar responded nicely")
* the rest of the GUI (SWT) was working ("THEN I pushed the green forward and back")

Out of curiosity, how "fast" are the progress bar updates? They're supposed to be once per second. Do they come faster (every half second) or slower (every 5 seconds). Are they consistent (update every second) or sporadic (progress bar updates, hangs for 2 seconds then continues)

I'm thinking that the event queue is being flooded, and may "decrease" the notification time from 1 second to 3 seconds (or more). Still, I'd like to see the log to get an idea about actual timing....

If you can try again and send the log, that would be great. If not, I'll just bump it up a few seconds (or add a config) and let you try again.

Thanks!

**Comment by markusmaresch** [comment]
Hi, 
now I have attached the second log file from this session - hope this helps. grep for xobc and look for "20160830_091649.281" and next "20160830_101230.249" - in between IT happened :-)  ... 

The bar updates were quite good - not super responsive, but acceptable - definitely faster than once every 2 or 5 seconds.

Other thought: The download really stopped - I now always have a Finder window (OS-X) open and am watching the respective directory tmp/bldr/xxxx.zip grow and update the partial zip and the end-marker file. Typically they grow at least once a minute (or up to 20 times a minute).
But when THIS happens, the download really stops. This is a reality - no growth of the 2 files.
Playing with the GUI, changing views to over windows and coming back somehow makes this continue.

Thanks again ! :-) 

[session.txt.zip](https://github.com/gnosygnu/xowa/files/447711/session.txt.zip)

**Comment by markusmaresch** [comment]
Hi,
one more update.
No change in base tree - still 3.5.8.1
Downloaded recent English update of articles + images (2016/08)
One temporary 503 error due to archive.org maintenance.
Rest was fully ok.
Logs below
:-) 

[20160831_160714.log.zip](https://github.com/gnosygnu/xowa/files/448296/20160831_160714.log.zip)
[session.txt.zip](https://github.com/gnosygnu/xowa/files/448298/session.txt.zip)

**Comment by gnosygnu** [comment]
Hi!

> now I have attached the second log file from this session - hope this helps. grep for xobc and look for "20160830_091649.281" and next "20160830_101230.249" - in between IT happened :-) ... 

Cool. I checked the log and gleaned the following:

* DownloadCentral was downloading between 8:50 and 9:16
* The notifications were fairly regular (every second)
* There is a stop, and then a resumption at 10:12. I guess this is when you noticed it and forced it to resume by "playing with the GUI"

Unfortunately, this throws out my "erratic updates" / "update flood" theory . But...

> The download really stopped - I now always have a Finder window (OS-X) open and am watching the respective directory tmp/bldr/xxxx.zip grow and update the partial zip and the end-marker file. Typically they grow at least once a minute (or up to 20 times a minute).
But when THIS happens, the download really stops. 

Great observation! Indeed, this means the download really stopped, which:
* Probably rules out the GUI since the download will continue even if the GUI hangs (the download runs on a separate non-GUI thread)
* Indicates that something failed at the downloading level (which is pretty low-level)

I looked thought the code, and realized I don't catch fatal download errors. I don't even log them. (Honestly, I thought I did both)

This means that you may have a network hiccup that killed the download. For example, I tried the following:
* Started a dewiki download in Download Central
* Let it download for a min (just to make sure stuff was working)
* Pulled the network plug
* The download just stopped. XOWA was still running
* Reloaded the page (slight sluggishness but not like what you described)
* Tried to resume. Could not
* Exited XOWA. Terminal was still running. I had to force-close it
* Restarted XOWA and resumed the download

So, it's possible that you had a brief outage and that this killed the XOWA download. I still can't get a crash report (nor the GUI unresponsiveness), but the behavior is very similar.

Let me add some logic to handle randomly severed connections (probably resume / retry for a few minutes). Also, I'll add better logging.

If you can, try to do the pull-network-cord test and see if you end up in the same state. If you do, then bingo: that is the issue. If not, then hopefully this patch will fix the issue. (by trying to resume downloading if it stops)

Will post again here when the patch is ready. Thanks!

**Comment by gnosygnu** [comment]
Hey. I updated the issue title just to provide more info. Let me know if that's not okay, and I'll restore the original version. (Stuck on (large) Download Central Job (in semaphore or /bin/bash ?))

So, I looked at this more, and discovered the following:
* There is code to handle (and log) fatal download errors. What happens is that the error message shows on the screen and the play button becomes changes from a pause button to either a retry button or a play button
* On Mac OS X, a download timeout doesn't throw an error. It just "hangs". I can pull the internet plug, and XOWA on Mac OS X will still think there's no error. Contrast this with Windows / Linux where an error will be thrown immediately.
* I can force Mac OS X to throw an error by setting an explicit timeout (connection.setReadTimeout). I'll add that to the next version.

So, I think the following happened:
* You're downloading a wiki from archive.org through Download Central.
* Some sort of download error occurs (Archive.org "flakes out". Or the connection gets dropped.)
* XOWA tries to fail immediately and report the error to the screen. However, for some reason, Mac OS X behaves differently than Windows and Linux and never officially fails.
* The next release of XOWA will do the following
** Force Mac OS X to fail by explicitly setting a setReadTimeout
** Try to be more resilient, by retrying a set number of times after a download failure

I'll post again here when the version is out. Otherwise, thanks for reporting the issue!

**Comment by markusmaresch** [comment]
I can test the following upon the next update - and not before next week end :-) 
        _The next release of XOWA will do the following
            Force Mac OS X to fail by explicitly setting a setReadTimeout
            Try to be more resilient, by retrying a set number of times after a download failure_

**Comment by gnosygnu** [comment]
Tonight's release (https://github.com/gnosygnu/xowa/releases/tag/v3.9.2.1) has a fix that does the following:

* If a connection is dropped, retry every 10 seconds 6 times
* If it still fails, report the error

I tested both on Mac OS X today with Simple Wikipedia. I forced a dropped connection by disabling wi-fi under System Preferences. It worked according to my expectations, so hopefully it will work according to yours. :)

Let me know if you run into any issues.

Thanks.
